### DISCUSSION TITLE
Const method parameter modifier

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#6771] [Language Ideas] [answered]
It would be very useful for libraries like Dapper or others that need high security if we could limit a method parameter to values known at compile time. This could be used to avoid things like SQL injection and make the intent of the method clearer. This could be used even for logging -- how many times have I seen people interpolate strings in their logs instead of using the overload with the params array? This could be avoided if the intent of the method was clearer, and the way I see it having the const modifier keyword usable in method parameters would make the intent very clear, and would be near impossible to mess up.

Examples:
```csharp
public static IEnumerable<T> Query<T>(const string sqlQuery, object parameters)
{
    // Handle SQL query. It is a lot harder for SQL injections to take place here.
}
```

```csharp
public static void Log(const string logFormat, params object[] parameters)
{
    // Handle logging. Now it is near impossible for the developer to mess up here.
}
```

## Worth noting:

- The const keyword would work exactly the same way it does when we declare fields or variables: the value must be known at compile time. Reference types other than string can only be null.

## Still not clear:

- Should this behavior only be enforced at compile-time? I believe so, but it is worth discussing.

**Comment by HaloFour** [comment]
The `const` keyword applies some additional behavior as well, namely that the member doesn't really exist and the name is replaced at the callsite by the value by the compiler.  A parameter couldn't be a `const` for that reason today.

I think the use case could be better served through analyzers which can key off of custom attributes and then inspect the code at the callsite to enforce specific business rules, like requiring it to be a literal.

**Comment by FaustVX** [comment]
You could use the String Interpolation Handler for this, with only `AppendLiteral` method in it.
The compiler will not compile with other that literal.
[sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgAwFgBQiCM65IARyYB0AStMAJYC2ApkQMID2VADhQDY1gDKXAbhQDGNAM4BudDkwA2AggKYA7OgDe6PBrwswFPgENgNBbLgAWPAFk9FKAAoAlOs1q0mt3gAyjAOYXRIvW8aWwASACJuAFEPSPoAFTwAKjwAMRIAeQs8YD0AI05MMPsJV3cNL19/QODwqJj4pNSMrJUwnPyaQoBfIpKy8p8/EQCg0Ijo2ITktMy8FSQe4qcNLuW8NbgAZmMCcwqhkeCARQguAE8ASShDMBZGdgMaABNuYB0obwAJPSgnzjA8CIAI7sE7nRylDQufp4fQAoGnMBnPAAXkBILBSKIHgoNz07D6mlWaGJ6AA2lcbncHoYXm8bF8fn8uABdHDbMA0ABmgLeEEEwDwmMu1y41MedPejN+/1UawA9PK8K9GGBqnguarsgALIy5CAcYAAWggLAUqEhBG2hCQAH5PLiuPi5kFgGItDp9IZATQ3XhSZatkLESKqfcJa8pd8ZVxbDZBexHWr2B4aB9gNqADR4eMa1VUAy0pjkCFuFT+ySB63mACCLBYaaeOLx7FsNsBpecazcFB5tmbTvYOZEeCgEHY7E7MPKSedaPE3c0NHYIhoi7KGbAjAA7qOaNuHISVugukA)

**Reply by 333fred** [reply]
Or, alternatively, just use the handlers in general. This is the type of scenario they excel at. You could have the `AppendFormatted` method escape any content passed to it, and you can use the conditional features to avoid evaluating anything if the current log level declares that it shouldn't be used.

**Reply by FaustVX** [reply]
Yes, it's better,
I've read only half of the question 😄

**Comment by Joe4evr** [comment, marked as answer]
You're in luck, .NET 7 just introduced [`[ConstantExpected]`](https://apisof.net/catalog/ea3f1012-685f-88fa-4071-b9bf8639ace2) which you can apply to method parameters so consumers get informed that they really should be passing a compile-time constant.

**Reply by ghost** [reply]
Thank you, @Joe4evr. Your answer was the only one so far that didn't feel like a hack or wasn't very practical. Awesome stuff! 😌👍

**Reply by 333fred** [reply]
I'm curious why you say that about interpolated string handlers. This is an intended use case for them.
